src/hotspot/share/opto/noOverflowInt.hpp

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright (c) 2024, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#ifndef SHARE_OPTO_NOOVERFLOWINT_HPP
+#define SHARE_OPTO_NOOVERFLOWINT_HPP
+
+#include "utilities/globalDefinitions.hpp"
+#include "utilities/ostream.hpp"
+
+// Wrapper around jint, which detects overflow.
+// If any operation overflows, then it returns a NaN.
+class NoOverflowInt {
+private:
+  bool _is_NaN; // overflow, uninitialized, etc.
+  jint _value;
+
+public:
+  // Default: NaN.
+  NoOverflowInt() : _is_NaN(true), _value(0) {}
+
+  // Create from jlong (or jint) -> NaN if overflows jint.
+  explicit NoOverflowInt(jlong value) : _is_NaN(true), _value(0) {
+    jint trunc = (jint)value;
+    if ((jlong)trunc == value) {
+      _is_NaN = false;
+      _value = trunc;
+    }
+  }
+
+  static NoOverflowInt make_NaN() { return NoOverflowInt(); }
```

Comment:
```suggestion
  static constexpr NoOverflowInt make_NaN() { return NoOverflowInt(); }
```
I think this can be constexpr, but you may need to add constexpr to the ctor as well.